private/bufpkg/bufimage/bufimageutil/image_index.go

```diff
@@ -16,110 +16,120 @@ package bufimageutil
+var _ namedDescriptor = (*descriptorpb.MethodDescriptorProto)(nil)
+
+type elementInfo struct {
+	fullName, file string
+	parent         namedDescriptor
 }
 
 // newImageIndexForImage builds an imageIndex for a given image.
 func newImageIndexForImage(image bufimage.Image, opts *imageFilterOptions) (*imageIndex, error) {
 	index := &imageIndex{
-		NameToDescriptor: make(map[string]protosource.NamedDescriptor),
+		ByName:       make(map[string]namedDescriptor),
+		ByDescriptor: make(map[namedDescriptor]elementInfo),
+		Files:        make(map[string]*descriptorpb.FileDescriptorProto),
 	}
 	if opts.includeCustomOptions {
-		index.NameToOptions = make(map[string]map[int32]protosource.Field)
+		index.NameToOptions = make(map[string]map[int32]*descriptorpb.FieldDescriptorProto)
 	}
 	if opts.includeKnownExtensions {
-		index.NameToExtensions = make(map[string][]protosource.Field)
+		index.NameToExtensions = make(map[string][]*descriptorpb.FieldDescriptorProto)
 	}
 
 	for _, file := range image.Files() {
-		// TODO: protosource.File is a heavyweight representation, and NewFile is not a cheap
-		//  operation. We only really need to gather fully-qualified names for each element
-		//  and to have some minimal level of upward references (e.g. access a descriptor's
-		//  parent element), which could be done with a far more efficient representation.
-		protosourceFile, err := protosource.NewFile(newInputFile(file))
-		if err != nil {
-			return nil, err
-		}
-		if err := protosource.ForEachMessage(func(message protosource.Message) error {
-			if storedDescriptor, ok := index.NameToDescriptor[message.FullName()]; ok && storedDescriptor != message {
-				return fmt.Errorf("duplicate for %q: %#v != %#v", message.FullName(), storedDescriptor, message)
+		fileName := file.Path()
+		fileDescriptorProto := file.Proto()
+		index.Files[fileName] = fileDescriptorProto
+		err := walk.DescriptorProtos(fileDescriptorProto, func(name protoreflect.FullName, msg proto.Message) error {
+			if _, ok := index.ByName[string(name)]; ok {
+				return fmt.Errorf("duplicate for %q", name)
 			}
-			index.NameToDescriptor[message.FullName()] = message
-			if !opts.includeKnownExtensions && !opts.includeCustomOptions {
-				return nil
+			descriptor, ok := msg.(namedDescriptor)
+			if !ok {
+				return fmt.Errorf("unexpected descriptor type %T", msg)
 			}
-			for _, field := range message.Extensions() {
-				index.NameToDescriptor[field.FullName()] = field
-				extendeeName := field.Extendee()
-				if opts.includeCustomOptions && isOptionsTypeName(extendeeName) {
-					if _, ok := index.NameToOptions[extendeeName]; !ok {
-						index.NameToOptions[extendeeName] = make(map[int32]protosource.Field)
-					}
-					index.NameToOptions[extendeeName][int32(field.Number())] = field
+			var parent namedDescriptor
+			if pos := strings.LastIndexByte(string(name), '.'); pos != -1 {
+				parent = index.ByName[string(name[:pos])]
+				if parent == nil {
```

Comment:
I don't quite follow the nit -- would you prefer this line also use the 2-value lookup, instead of the nil check? I did them differently primarily because they are fundamentally different: the one above is just doing an "exists" check to see if an element is already in the map. But here, we need the actual value from the map.

Reply:
I went ahead and made them consistent, both check the value against nil. If you'd prefer to see them instead key off the `ok` of a 2-value map query, I don't mind making that change (6 vs. 1/2 dozen to me 🤷).